saw-core/src/SAWCore/Term/CtxTerm.hs

```diff
@@ -1024,60 +632,57 @@ ctxReduceRecursor rec elimf c_args CtorArgStruct{..} =
 --   iota reduction for @ctxReduceRecursor@.  We assume
 --   the input terms we are given live in an ambient
 --   context @amb@.
-ctxReduceRecursor_ :: forall m amb d ixs elim args.
+ctxReduceRecursor_ :: forall m.
   MonadTerm m =>
-  CtxTerm  amb (Rec d) {- ^ recursor value eliminatiting data type d -}->
-  CtxTerm  amb elim    {- ^ eliminator function for the constructor -} ->
-  CtxTerms amb args    {- ^ constructor actual arguments -} ->
-  Bindings (CtorArg d ixs) amb args
+  Term     {- ^ recursor value eliminatiting data type d -}->
+  Term     {- ^ eliminator function for the constructor -} ->
+  [Term]    {- ^ constructor actual arguments -} ->
+  [(LocalName, CtorArg)]
     {- ^ telescope describing the constructor arguments -} ->
   m Term
 ctxReduceRecursor_ rec fi args0 argCtx =
-  do args <- mk_args CtxTermsCtxNil args0 argCtx
-     whnfTerm =<< foldM (\f arg -> mkTermF $ App f arg) (unAmb fi) args
+  do args <- mk_args [] args0 argCtx
+     whnfTerm =<< foldM (\f arg -> mkTermF $ App f arg) fi args
 
  where
-   -- extract a raw term into the ambient context
-    unAmb :: forall tp. CtxTerm amb tp -> Term
-    unAmb (CtxTerm t) = t
-
-    mk_args :: forall ctx xs.
-               CtxTermsCtx amb ctx ->  -- already processed parameters/arguments
-               CtxTerms    amb xs ->   -- remaining actual arguments to process
-               Bindings (CtorArg d ixs) (amb<+>ctx) xs ->
+    mk_args :: [Term] ->  -- already processed parameters/arguments
+               [Term] ->     -- remaining actual arguments to process
+               [(LocalName, CtorArg)] ->
                  -- telescope for typing the actual arguments
                m [Term]
     -- no more arguments to process
-    mk_args _ _ NoBind = return []
+    mk_args _ _ [] = return []
 
     -- process an argument that is not a recursive call
-    mk_args pre_xs (CtxTermsCons x xs) (Bind _ (ConstArg _) args) =
-      do tl <- mk_args (CtxTermsCtxCons pre_xs x) xs args
-         pure (unAmb x : tl)
+    mk_args pre_xs (x : xs) ((_, ConstArg _) : args) =
+      do tl <- mk_args (pre_xs ++ [x]) xs args
+         pure (x : tl)
 
     -- process an argument that is a recursive call
-    mk_args pre_xs (CtxTermsCons x xs) (Bind _ (RecursiveArg zs ixs) args) =
-      do zs'  <- ctxSubstInBindings pre_xs InvNoBind NoBind zs
-         ixs' <- ctxSubstInBindings pre_xs InvNoBind zs ixs
+    mk_args pre_xs (x : xs) ((_, RecursiveArg zs ixs) : args) =
+      do zs'  <- ctxSubstInBindings pre_xs [] [] zs
+         ixs' <- ctxSubstInBindings pre_xs [] zs ixs
          recx <- mk_rec_arg zs' ixs' x
-         tl   <- mk_args (CtxTermsCtxCons pre_xs x) xs args
-         pure (unAmb x : recx : tl)
+         tl   <- mk_args (pre_xs ++ [x]) xs args
+         pure (x : recx : tl)
+
+    mk_args _ _ _ = error "mk_args: impossible"
```

Comment:
Why is this case impossible to reach? Is there a precondition about the lengths of the arguments that supports this claim? If so, it would be worth documenting this precondition in the comments.

Reply:
Fixed in 865e96e790481797adf3f2e5b4bb1e270f0e0f91.